### PR TITLE
feat: Claude Code Wrapped — per-project weekly digest emails

### PR DESCRIPTION
**Pull request by csansoon**
A "Spotify Wrapped"-style weekly summary email for any Latitude project receiving Claude Code telemetry (spans carrying `metadata['claude_code.version']`). One email per project, sent to every verified-email member of the project's organization, gated by an org-level feature flag.

## triggers

Two paths, both publishing the same `runForProject` task on the new `claude-code-wrapped` queue topic:

- **Friday 09:00 UTC cron**, registered at worker boot via BullMQ's `upsertJobScheduler`. The handler computes `[now-7d, now]`, lists projects with Claude Code activity in that window (ClickHouse), intersects with orgs that have the `claude-code-wrapped` feature flag enabled (admin postgres), and fans out one task per surviving `(orgId, projectId)`.
- **Manual** trigger button on `/backoffice/projects/<id>` for staff. Same task, same window, same gate.

## the pipeline

`runClaudeCodeWrappedUseCase` per project:

1. Re-checks the feature flag (defense-in-depth — the cron pre-filters, the manual button doesn't).
2. Counts CC sessions in the window. Zero → skipped, no email.
3. Loads org members; filters by `emailVerified`.
4. Calls `buildReportUseCase` — eight ClickHouse queries in parallel (concurrency 6) plus a per-top-3-workspace fan-out for deep-dive data.
5. Renders + sends the email per recipient (concurrency 5).

## what's in the email

Top-to-bottom: hero (project name, UTC week range), 4-card headline grid (sessions / Claude time / files touched / commands), breadth strip with streak + test-run callout, a big "lines written" reveal with a scaling comparison anchor (`"≈ 10% of the Apollo 11 guidance code"`, `"≈ Doom"`, …), read/write ratio with a Library-of-Congress-style anchor for read volume, 7×24 hour-of-day heatmap (UTC), three moment cards (longest session · busiest day · biggest single write), the user's #1 bash command, one section per top-3 workspace (top files with `+N / −M` or `N reads`, top 3 bash commands, active branches), and a personality reveal card with a PNG illustration.

Six personality archetypes assigned by a pure priority cascade in `assignPersonality`: Strategist (planning-heavy) > Marathoner (long avg session) > tool-mix winner among Surgeon (edit) / Architect (write) / Detective (read+search) / Conductor (bash).

## new concepts

- **`claude-code-wrapped` queue topic** with two tasks: `triggerWeeklyRun` (cron) and `runForProject` (per-project).
- **`QueuePublisherShape.scheduleRepeatable`** — first BullMQ repeatable-job helper in the codebase. Wraps `Queue.upsertJobScheduler` so registration is idempotent across worker restarts.
- **`AdminFeatureFlagRepository.findEligibilityForFlag`** — returns `{ enabledForAll, organizationIds }` so cross-org workers can scope their workload without enumerating all orgs.
- **`CLAUDE_CODE_WRAPPED_FLAG`** typed constant in `@domain/feature-flags`. The flag itself is created at runtime through `/backoffice/feature-flags/` — no code-side registry.
- **`ClaudeCodeSpanReader`** port + ClickHouse adapter — 12 read methods covering totals, session durations, LOC stats, tool mix, top files, top bash commands, top workspaces, top branches, per-workspace deep-dives (with per-file `+N / −M` sums and top-3 commands via CTEs), heatmap, busiest day, biggest single write.
- **`Report` Zod entity** in `@domain/spans/claude-code-wrapped/entities/report.ts` — the data shape every template field reads from. Zero computation in the template.
- **`@domain/email/components/WrappedLayout`** + **`EmailHeading`** — sibling to the existing `ContainerLayout`, with Anthropic-style tokens (`#D97555` accent on `#F0EEE6` cream, Georgia serif) scoped under `emailDesignTokens.colors.claude.*`.
- **Personality PNGs** at `apps/web/public/email-branding/claude-code-wrapped/personalities/{surgeon,architect,…}.png`, served by the web app at `${LAT_WEB_URL}/…`.

## design decisions worth knowing

- **Per-project, not per-user.** Each project's email goes to every verified member of the project's org. No per-user dedupe.
- **No per-user opt-out.** The org-level feature flag is the only gate. Per-user preferences are deferred until Wrapped graduates from flag-gated.
- **Privacy on paths.** File paths in workspace cards are rendered relative to `metadata['workspace.path']` (e.g. `src/Chat.tsx`, never `/Users/<name>/Dev/.../Chat.tsx`). Files outside the workspace fall back to basename. The biggest-write moment shows basename only.
- **LOC = newline counts** across `tool_input.content` (Write), `new_string` / `old_string` (Edit / NotebookEdit, plus MultiEdit's `edits[]`), and `tool_output` (Read / NotebookRead). Approximate but a single ClickHouse pass.
- **Heatmap is UTC**, called out in a caption. Per-user timezone is a follow-up once any TZ source exists.
- **`runForProject` has no BullMQ dedup key.** A previous version used a date-keyed dedup; failed jobs kept the jobId "burned" and blocked manual retries. The cron only fires weekly and manual clicks are staff-driven, so duplicate-publish risk is low. Comment in the worker.
- **Image base URLs are env-aware.** Worker reads `LAT_WEB_URL` (same env Better Auth and file-uploads use) at startup and bakes it into the `renderEmail` closure. The template derives personality images, project deep-link, settings link, and logo from that single base.
- **Personality algorithm uses absolute shares**, not population-relative. Known limitation — most users will land in Detective. Recalibrating with population data is a follow-up.
- **Email footer carries a project deep-link** (`/projects/<slug>`) and an unsubscribe placeholder pointing at `/settings/account`. The settings page doesn't host a per-user toggle yet; the link is wired up for when it does.

## tests

- **Unit (491 in `@domain/spans`)**: anchor picker ramps, personality assignment (6 archetypes × tie-break + edge cases), `assembleReport` (bucketisation, heatmap zero-fill, relative-path stripping, per-file diff carry-through, top-3 commands, workspace-count branching).
- **Worker integration (6 in `apps/workers`)**: `fanOutWeeklyRunUseCase` covers the "(has spans) ∩ (flag enabled)" intersection — empty, no-eligible-orgs, partial overlap, `enabledForAll: true`, flagged-org-with-no-spans, window propagation.
- **Adapter (PGlite)**: `findEligibilityForFlag` covered in `admin-feature-flag-repository.test.ts`.
- **No ClickHouse adapter tests.** The repo has no in-process ClickHouse testkit; the reader is verified end-to-end via the backoffice trigger + Mailpit.

## deliberately deferred

- Per-user opt-out / email preferences table.
- Per-organization timezone for the heatmap.
- Population-relative personality scoring.
- Bash sub-command grouping (`git commit` vs `git log`).
- A "preview" mode that runs the analysis without sending.

---

For reviewers: the personality reveal and per-workspace sections are the most visually interesting bits — recommend triggering the manual flow against a real project and inspecting in Mailpit before sign-off.

### CODE REVIEW

File: packages/domain/spans/src/claude-code-wrapped/use-cases/build-report.ts [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+  const mix = emptyToolMix()
+  for (const row of rows) {
+    mix[toolBucketFor(row.toolName)] += row.uses
+  }
+  return mix
+}
+
+const basename = (path: string): string => {
+  if (path === "") return ""
+  const trimmed = path.replace(/\/+$/, "")
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/claude-code-wrapped/use-cases/build-report.ts]: Addressed in 285bb22f0. Replaced the trailing-slash regex with a linear charCodeAt loop over the path — no backtracking, behaviour unchanged for the inputs we actually see (POSIX / and Windows \\). Same alert should clear on re-scan.

File: apps/web/src/domains/admin/claude-code-wrapped.functions.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  projectId: z.string().min(1).max(256),
+})
+
+/**
+ * Manually trigger the Claude Code Wrapped pipeline for a single project.
+ *
+ * Looks up the project's organization (so the client can't claim an orgId it
+ * doesn't own), then publishes a `runForProject` task on the
+ * `claude-code-wrapped` topic with a date-keyed dedupe so multiple clicks in
+ * the same day collapse into one job.
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/domains/admin/claude-code-wrapped.functions.ts]: Good catch — the docstring was stale. Updated in 94b023db1 to match the actual behaviour (no `dedupeKey` on this publish; each click is its own job). The comment now explains why: an earlier date-keyed dedup left failed jobIds 'burned' in BullMQ and blocked legitimate retries, so we dropped it.

File: packages/domain/spans/src/claude-code-wrapped/use-cases/build-report.ts [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    readonly row: WorkspaceDeepDiveRow
+  }>
+}
+
+/**
+ * Pure assembly step — no Effect, no I/O. Bucketises the tool mix, computes
+ * relative paths + anchors, fills the 7×24 heatmap, assigns the personality,
+ * and validates the result against `reportSchema` (callers can rely on the
+ * returned object being a `Report`).
+ */
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/claude-code-wrapped/use-cases/build-report.ts:173]: Fixed in 9ad7c6363 — went with the validation rather than weakening the contract. The function now wraps the assembled object in `reportSchema.parse(...)` so the comment is honest and any future drift between assembly and schema (e.g. a forgotten field after a query refactor) trips at the seam.

File: packages/domain/spans/src/claude-code-wrapped/use-cases/assign-personality.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    readonly buildEvidence: () => readonly [string, string, string]
+  }> = [
+    {
+      kind: "surgeon",
+      share: shareOf("edit"),
+      buildEvidence: () => [
+        `${formatPercent(shareOf("edit"))} of your tool calls were Edits`,
+        `Touched ${filesTouched.toLocaleString("en-US")} file${filesTouched === 1 ? "" : "s"} this week`,
+        `${toolMix.write.toLocaleString("en-US")} new file${toolMix.write === 1 ? "" : "s"} written from scratch`,
+      ],
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/claude-code-wrapped/use-cases/assign-personality.ts:122]: Agreed — Write overwrites the path, doesn't necessarily create it. Reworded in 757e090dc to drop the 'new file from scratch' claim — the line now says 'N Write call(s) on top'.

File: packages/domain/spans/src/claude-code-wrapped/use-cases/assign-personality.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    },
+    {
+      kind: "architect",
+      share: shareOf("write"),
+      buildEvidence: () => [
+        `${formatPercent(shareOf("write"))} of your tool calls created new files`,
+        `${toolMix.write.toLocaleString("en-US")} Write call${toolMix.write === 1 ? "" : "s"}`,
+        `Touched ${filesTouched.toLocaleString("en-US")} file${filesTouched === 1 ? "" : "s"} in total`,
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/claude-code-wrapped/use-cases/assign-personality.ts:130]: Same fix as the Surgeon thread, in 757e090dc. Architect's evidence now reads 'X% of your tool calls were Writes / N file(s) written' — no 'new' claim.

File: packages/domain/email/src/templates/claude-code-wrapped/-components/WorkspaceSection.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+          <table
+            cellPadding={0}
+            cellSpacing={0}
+            border={0}
+            role="presentation"
+            style={{ width: "100%", borderCollapse: "collapse" }}
+          >
+            <tbody>
+              {topFiles.map((file, idx) => (
+                <tr key={file.displayPath} style={idx < topFiles.length - 1 ? rowDividerStyle : undefined}>
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/email/src/templates/claude-code-wrapped/-components/WorkspaceSection.tsx]: Fixed in 1b06a693d. Key is now `${idx}-${file.displayPath}` so two files collapsing to the same display path (e.g. workspacePath unknown + matching basenames) still get unique React keys.

**Review by andresgutgon — APPROVED** [review]
